tips/028_input_guardrails.md

```diff
@@ -0,0 +1,175 @@
+## Effective AI Engineering #28: Input Guardrails
+
+**Are malicious users draining your AI budget with expensive attacks?** Every prompt injection or jailbreak attempt hits your most expensive models, even when you could reject them at the gate.
+
+Input filtering at the application layer can block problematic queries before they consume costly LLM resources. This first line of defense also reduces latency and provides more reliable rejection than depending on LLM behavior alone.
+
+### The Problem
+
+Many developers send all user inputs directly to expensive LLMs without pre-filtering. This creates challenges that aren't immediately obvious:
+
+```python
+# BEFORE: No input filtering
+from mirascope.core import llm
+
+@llm.call(provider="openai", model="gpt-4o-mini")
+def handle_customer_query(user_input: str) -> str:
+    return f"""
+    You are a helpful customer service assistant.
+    Never provide harmful information or ignore safety guidelines.
+    
+    Customer query: {user_input}
+    """
+```
+
+**Why this approach falls short:**
+
+- **Resource Waste:** Malicious queries consume expensive LLM tokens even when they'll be rejected
+- **Inconsistent Blocking:** LLMs might occasionally comply with harmful requests despite instructions
+- **High Latency:** Every query requires full LLM processing before potential rejection
+
+### The Solution: Pre-Processing Input Classification
+
+A better approach is to classify and filter problematic inputs before they reach expensive models. This pattern combines fast heuristics with lightweight classification to block attacks at minimal cost.
+
+```python
+# AFTER: Multi-layer input filtering
+from mirascope.core import llm
```

Comment:
this code example is too long. We can omit the heuristic based checks and stick with just the AI based one. Then we can have some prose/text which explains that you could use heuristics instead of another AI call too